In [6]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
""
#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [7]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df=application_df.drop(columns=['EIN','NAME'])
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [9]:
# Look at APPLICATION_TYPE value counts for binning
vc1 = application_df['APPLICATION_TYPE'].value_counts()

# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
#Cutoff Value of 100

application_types_to_replace = vc1[vc1<100]
# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [10]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#class_list = application_df["CLASSIFICATION"].unique()[5:]
vc = application_df.CLASSIFICATION.value_counts()
vc[vc>1]

# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = vc[vc<1000].index.tolist()
classifications_to_replace
# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [11]:
# Convert categorical data to numeric with `pd.get_dummies`
new_df =pd.get_dummies(application_df)
new_df

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  \
0           1      5000              1                     1   
1           1    108590              1                     0   
2           1      5000              0                     0   
3           1      6692              1                     0   
4           1    142590              1                     0   
...       ...       ...            ...                   ...   
34294       1      5000              0                     0   
34295       1      5000              0                     0   
34296       1      5000              0                     0   
34297       1      5000              1                     0   
34298       1  36500179              0                     0   

       APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  \
0                         0                     0                     0   
1                         0                     0                     0   
2                         0                     0                     0   
3                         0                     0                     0   
4                         0                     0                     0   
...                     ...                   ...                   ...   
34294                     0                     0                     0   
34295                     0                     0                     0   
34296                     0                     0                     0   
34297                     0                     0                     0   
34298                     0                     0                     0   

       APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  APPLICATION_TYPE_T19  ...  \
0                         0                     0                     0  ...   
1                         0                     0                     0  ...   
2                         0                     0                     0  ...   
3                         0                     0                     0  ...   
4                         0                     0                     0  ...   
...                     ...                   ...                   ...  ...   
34294                     0                     0                     0  ...   
34295                     0                     0                     0  ...   
34296                     0                     0                     0  ...   
34297                     0                     0                     0  ...   
34298                     0                     0                     0  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                      0                       0                         0   
1                      1                       0                         0   
2                      0                       0                         0   
3                      0                       1                         0   
4                      0                       0                         1   
...                  ...                     ...                       ...   
34294                  0                       0                         0   
34295                  0                       0                         0   
34296                  0                       0                         0   
34297                  0                       0                         0   
34298                  0                       0                         0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                       0                 0                       0   
1                       0                 0                       0   
2                       0                 0                       0   
3                       0                 0                       0   
4                       0                 0                       0   
...          

In [12]:
# Split our preprocessed data into our features and target arrays
y=new_df.IS_SUCCESSFUL.values
X=new_df.drop(columns="IS_SUCCESSFUL")

# Split the preprocessed data into a training and testing dataset
X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=42, stratify=y)

In [13]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

ORIGINAL MODEL

In [44]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn4 = tf.keras.models.Sequential()




# First hidden layer
nn4.add(tf.keras.layers.Dense(units=16, activation="relu", input_dim=51))

# Second hidden layer
nn4.add(tf.keras.layers.Dense(units=16, activation="relu"))
# Output layer
nn4.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn4.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_23 (Dense)            (None, 16)                832       
                                                                 
 dense_24 (Dense)            (None, 16)                272       
                                                                 
 dense_25 (Dense)            (None, 1)                 17        
                                                                 
Total params: 1,121
Trainable params: 1,121
Non-trainable params: 0
_________________________________________________________________


In [45]:
nn4.compile(loss="binary_crossentropy",optimizer = "adam",metrics=["accuracy"])

In [46]:
fit_model = nn4.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5845 - accuracy: 0.7084
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5571 - accuracy: 0.7309
Epoch 3/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5530 - accuracy: 0.7322
Epoch 4/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5516 - accuracy: 0.7314
Epoch 5/100
804/804 [==============================] - 4s 5ms/step - loss: 0.5500 - accuracy: 0.7325
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5490 - accuracy: 0.7336
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5487 - accuracy: 0.7337
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5480 - accuracy: 0.7348
Epoch 9/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5474 - accuracy: 0.7342
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5470 - accura

In [47]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

# Evaluate the model using the test data
model_loss, model_accuracy = nn4.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss2}, Accuracy: {model_accuracy2}")

268/268 - 0s - loss: 0.5553 - accuracy: 0.7247 - 463ms/epoch - 2ms/step
Loss: 0.5564376711845398, Accuracy: 0.7250145673751831


8 units per inputlayer - Loss: 0.556532621383667, Accuracy: 0.7239649891853333
16 units per input layerr- Accuracy: 0.7250145673751831


Model Optimized by doubling epochs to 200

In [18]:
fit_model_200 = nn.fit(X_train_scaled, y_train, epochs=200)

Epoch 1/200
804/804 [==============================] - 3s 4ms/step - loss: 0.5426 - accuracy: 0.7367
Epoch 2/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5428 - accuracy: 0.7360
Epoch 3/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5427 - accuracy: 0.7373
Epoch 4/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5425 - accuracy: 0.7367
Epoch 5/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5428 - accuracy: 0.7365
Epoch 6/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5427 - accuracy: 0.7370
Epoch 7/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5426 - accuracy: 0.7376
Epoch 8/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5427 - accuracy: 0.7363
Epoch 9/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5426 - accuracy: 0.7371
Epoch 10/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5424 - accura

In [19]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5563 - accuracy: 0.7219 - 551ms/epoch - 2ms/step
Loss: 0.5562570095062256, Accuracy: 0.7218658924102783


Loss: 0.5562570095062256, Accuracy: 0.7218658924102783

Optimization Attempt #2 Adding more hidden layers

In [26]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn2 = tf.keras.models.Sequential()




# First hidden layer
nn2.add(tf.keras.layers.Dense(units=8, activation="relu", input_dim=51))

# Second hidden layer
nn2.add(tf.keras.layers.Dense(units=8, activation="relu"))
# Third hidden layer
nn2.add(tf.keras.layers.Dense(units=8, activation="relu"))
# Output layer
nn2.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn2.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_13 (Dense)            (None, 8)                 416       
                                                                 
 dense_14 (Dense)            (None, 8)                 72        
                                                                 
 dense_15 (Dense)            (None, 8)                 72        
                                                                 
 dense_16 (Dense)            (None, 1)                 9         
                                                                 
Total params: 569
Trainable params: 569
Non-trainable params: 0
_________________________________________________________________


In [27]:
nn2.compile(loss="binary_crossentropy",optimizer = "adam",metrics=["accuracy"])

In [34]:
fit_model = nn2.fit(X_train_scaled, y_train, epochs=500)

Epoch 1/500
804/804 [==============================] - 2s 3ms/step - loss: 0.5422 - accuracy: 0.7367
Epoch 2/500
804/804 [==============================] - 2s 2ms/step - loss: 0.5421 - accuracy: 0.7372
Epoch 3/500
804/804 [==============================] - 2s 2ms/step - loss: 0.5420 - accuracy: 0.7369
Epoch 4/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5422 - accuracy: 0.7356
Epoch 5/500
804/804 [==============================] - 2s 2ms/step - loss: 0.5421 - accuracy: 0.7369
Epoch 6/500
804/804 [==============================] - 2s 2ms/step - loss: 0.5422 - accuracy: 0.7361
Epoch 7/500
804/804 [==============================] - 2s 2ms/step - loss: 0.5422 - accuracy: 0.7369
Epoch 8/500
804/804 [==============================] - 2s 2ms/step - loss: 0.5423 - accuracy: 0.7372
Epoch 9/500
804/804 [==============================] - 2s 3ms/step - loss: 0.5415 - accuracy: 0.7370
Epoch 10/500
804/804 [==============================] - 2s 2ms/step - loss: 0.5421 - accura

In [37]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5582 - accuracy: 0.7251 - 736ms/epoch - 3ms/step
Loss: 0.5582181811332703, Accuracy: 0.7251312136650085


#Adding a hidden layer slightly improved the models accuracy.


Optimation Attempt #3: Changing the activation function of the second layer
Loss: 0.5564295649528503, Accuracy: 0.7252478003501892

In [30]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn3 = tf.keras.models.Sequential()




# First hidden layer
nn3.add(tf.keras.layers.Dense(units=4, activation="relu", input_dim=51))

# Second hidden layer
nn3.add(tf.keras.layers.Dense(units=4, activation="tanh"))
# Output layer
nn3.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn3.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_17 (Dense)            (None, 4)                 208       
                                                                 
 dense_18 (Dense)            (None, 4)                 20        
                                                                 
 dense_19 (Dense)            (None, 1)                 5         
                                                                 
Total params: 233
Trainable params: 233
Non-trainable params: 0
_________________________________________________________________


In [31]:
nn3.compile(loss="binary_crossentropy",optimizer = "adam",metrics=["accuracy"])

In [32]:
fit_model = nn3.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 6s 5ms/step - loss: 0.6120 - accuracy: 0.6774
Epoch 2/100
804/804 [==============================] - 4s 5ms/step - loss: 0.5709 - accuracy: 0.7232
Epoch 3/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5645 - accuracy: 0.7273
Epoch 4/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5600 - accuracy: 0.7297
Epoch 5/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5581 - accuracy: 0.7297
Epoch 6/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5565 - accuracy: 0.7299
Epoch 7/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5557 - accuracy: 0.7289
Epoch 8/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5549 - accuracy: 0.7304
Epoch 9/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5544 - accuracy: 0.7297
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5541 - accura

In [33]:
model_loss, model_accuracy = nn3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5593 - accuracy: 0.7234 - 923ms/epoch - 3ms/step
Loss: 0.5592879056930542, Accuracy: 0.7233819365501404


Optimization attempt 3 has a small decrease in accuracy